From ec8e573d64659c49c7bb45f00fdcce27a3a9efe0 Mon Sep 17 00:00:00 2001
From: Billy Laws <blaws05@gmail.com>
Date: Mon, 1 Aug 2022 00:03:36 -0400
Subject: [PATCH] inputflinger: filter inputs of disabled devices

Some controllers like the Nintendo JoyCons should only have
inputs read when enabled, allowing vendor-provided configs to
disable the controller under various conditions (i.e. only one
JoyCon is connected)

Change-Id: I4aa044daf530e64e927a86f40585a287ada7922b
---
 services/inputflinger/reader/EventHub.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/services/inputflinger/reader/EventHub.cpp b/services/inputflinger/reader/EventHub.cpp
index a1514af668..cb3ba7bb48 100644
--- a/services/inputflinger/reader/EventHub.cpp
+++ b/services/inputflinger/reader/EventHub.cpp
@@ -1305,6 +1305,15 @@ status_t EventHub::openDeviceLocked(const char* devicePath) {
     // Load the configuration file for the device.
     loadConfigurationLocked(device);
 
+    bool deviceDisabled = false;
+    if (device->configuration && device->configuration->tryGetProperty(String8("device.disabled"), deviceDisabled)) {
+        if (deviceDisabled) {
+            delete device;
+            ALOGV("ignoring disabled device");
+            return -1;
+        }
+    }
+
     // Figure out the kinds of events the device reports.
     ioctl(fd, EVIOCGBIT(EV_KEY, sizeof(device->keyBitmask)), device->keyBitmask);
     ioctl(fd, EVIOCGBIT(EV_ABS, sizeof(device->absBitmask)), device->absBitmask);
-- 
2.37.1

